zebra-consensus/src/primitives/groth16.rs

```diff
@@ -1,55 +1,230 @@
+//! Async Groth16 batch verifier service
+
 use std::{
+    fmt,
     future::Future,
+    mem,
     pin::Pin,
     task::{Context, Poll},
 };
 
+use bellman::{
+    groth16::{PreparedVerifyingKey, Proof},
+    VerificationError,
+};
+use bls12_381::Bls12;
+use pairing::MultiMillerLoop;
+use rand::{thread_rng, CryptoRng, RngCore};
+use tokio::sync::broadcast::{channel, error::RecvError, Sender};
 use tower::Service;
-
-use zebra_chain::primitives::Groth16Proof;
+use tower_batch::BatchControl;
+use tower_fallback::Fallback;
 
 use crate::BoxError;
 
-/// Provides verification of Groth16 proofs for a specific statement.
-///
-/// Groth16 proofs require a proof verification key; the [`Verifier`] type is
-/// responsible for ownership of the PVK.
-pub struct Verifier {
-    // XXX this needs to hold on to a verification key
+/// A Groth16 verification item, used as the request type of the service.
+// pub type Item = batch::Item<Bls12>;
+
+#[derive(Clone)]
+pub struct Item<E: MultiMillerLoop> {
```

Comment:
You might want to delete the commented-out type alias, and use the doc comment for the current `Item` type:
```suggestion
/// A Groth16 verification item, used as the request type of the service.
#[derive(Clone)]
pub struct Item<E: MultiMillerLoop> {
```

Reply:
The type alias should stay, since the `Item` is a filler for the item type from `bellman`, once the batch API has been completed, so all the comments left are the code that is supposed to be ripped out for the real implementation once the crypto side of things is finished. I'll go ahead and move all the temporary filler types to one part of the file and document why they're there and how they should be handled, but I'd like to leave in these commented lines because they should already be correct, and only need to be uncommented once we swap to the new bellman.

Reply:
Sounds good.

Sometimes there's a bit of a delay between follow-up PRs, so it's good to have a comment that references the follow-up.